### PR TITLE
Docs: add note about CMD and ENTRYPOINT commands

### LINKED ISSUE #10084: Intersection of CMD and ENTRYPOINT is ill-defined
I ran an experiment today to see how these things intersect.

https://gist.github.com/thockin/c44a88b84c2a1041a1bd

There are three cells in this matrix that are totally nonsensical.  Worse the dockerfile spec says "The shell form prevents any CMD or run command line arguments from being used", which is clearly untrue.

More that this, though, the semantic between the two is sort of weird, especially since they can be arrays or strings.

I wonder if there has been any thought put into simplifying the semantics around these things?

### PR DESCRIPTION
**Pull request by tkopczynski**
Adding a note in Dockerfile reference docs about CMD and ENTRYPOINT relation.

Fixes #10084 

Signed-off-by: Tomasz Kopczynski tomek@kopczynski.net.pl

### CODE REVIEW

File: docs/reference/builder.md [hunk 1]
```diff
@@ -1244,6 +1244,29 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
...
+
+Table below describes what command will be executed depending on `ENTRYPOINT` / `CMD` settings:
+
+|                                | No ENTRYPOINT              | ENTRYPOINT exec_entry p1_entry                            | ENTRYPOINT ["exec_entry", "p1_entry"]          |
+|--------------------------------|----------------------------|-----------------------------------------------------------|------------------------------------------------|
+| **No CMD**                     | *error, not allowed*       | /bin/sh -c exec_entry p1_entry                            | exec_entry p1_entry                            |
+| **CMD ["exec_cmd", "p1_cmd"]** | exec_cmd p1_cmd            | /bin/sh -c exec_entry p1_entry exec_cmd p1_cmd            | exec_entry p1_entry exec_cmd p1_cmd            |
+| **CMD ["p1_cmd", "p2_cmd"]**   | p1_cmd p2_cmd              | /bin/sh -c exec_entry p1_entry p1_cmd p2_cmd              | exec_entry p1_entry p1_cmd p2_cmd              |
+| **CMD exec_cmd p1_cmd**        | /bin/sh -c exec_cmd p1_cmd | /bin/sh -c exec_entry p1_entry /bin/sh -c exec_cmd p1_cmd | exec_entry p1_entry /bin/sh -c exec_cmd p1_cmd |
```
Comment (MHBauer) [inline docs/reference/builder.md]: I think a lot of this table probably needs to formatted backtick verbatim.

File: docs/reference/builder.md [hunk 2]
```diff
@@ -1244,6 +1244,29 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
 This signal can be a valid unsigned number that matches a position in the kernel's syscall table, for instance 9,
 or a signal name in the format SIGNAME, for instance SIGKILL.
 
+## CMD and ENTRYPOINT relation
```
Comment (MHBauer) [inline docs/reference/builder.md]: not sure I personally like the word relation. maybe `interaction`? `relationship`? something else? maybe it's fine.

**Comment by MHBauer** [comment]
formatting and table look good in `make docs`. not sure anything needs to change, but let's wait for more review.

@thaJeztah for review

File: docs/reference/builder.md (hunk 2 above)
Comment (thaJeztah) [inline docs/reference/builder.md]: "Understand how CMD and ENTRYPOINT interact" ? Better suggestions are welcome

File: docs/reference/builder.md (hunk 1 above)
Comment (thaJeztah) [inline docs/reference/builder.md]: :+1: agreed

File: docs/reference/builder.md [hunk 3]
```diff
@@ -1244,6 +1244,29 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
...
+1. Dockerfile should specify at least one of `CMD` or `ENTRYPOINT` commands.
+
+2. `ENTRYPOINT` should be defined when using the container as an executable.
+
+3. `CMD` should be used as a way of defining default arguments for an `ENTRYPOINT` command
+or for executing an ad-hoc command in a container.
+
+4. `CMD` will be overridden when running the container with alternative arguments.
+
+Table below describes what command will be executed depending on `ENTRYPOINT` / `CMD` settings:
```
Comment (thaJeztah) [inline docs/reference/builder.md]: Perhaps
```Markdown
The table below shows what command is executed for different `ENTRYPOINT` / `CMD` combinations: 
```

**Comment by tkopczynski** [comment]
@MHBauer @thaJeztah I've changed the lines you were not happy about. 
I left the table untouched because when using backticks every cell should have them, I'm not sure if it would look better.

What do you think?

**Comment by thaJeztah** [comment]
Hi @tkopczynski I just built the documentation from this PR (using `make docs`), looking at it, I'm thinking we should move this section below `ENTRYPOINT` (after http://docker.local:8000/engine/reference/builder/#shell-form-entrypoint-example), and change the heading to a `H3` (`###`). wdyt?

While I still think the examples in the table should be in backticks, I think we should hold that off for now, because the current layout of the documentation does not have enough space to fit the table in that case;

Without backticks;
before

With backticks;
after

So I'm okay for now to leave the backticks out.

**Comment by tkopczynski** [comment]
@thaJeztah I was not sure about the position of this note either, so I moved it to the ENTRYPOINT section as you suggested and it does look better :)

I agree that this table looks better with backticks (commands are more readable) but unfortunately for some reason I don't get a scrollbar in my browser (13-inch screen, Chrome) when page gets too wide. So  probably the current design is not ready for such input indeed ;)

**Comment by thaJeztah** [comment]
@tkopczynski thanks! Let's keep it like this for now, we can revisit in future; the information is there, which is good.

LGTM

**Comment by thaJeztah** [comment]
ping @MHBauer ptal

**Comment by MHBauer** [comment]
LGTM.

In the context I think it is fine.

[jokingly] why not rotate the table 90 degrees? go vertical!

**Comment by thaJeztah** [comment]
> [jokingly] why not rotate the table 90 degrees? go vertical!

There you go:
```
<style> table { transform: rotate(90.0deg) } </style>
```

**Pull request merged**